waimak-core/src/main/scala/com/coxautodata/waimak/dataflow/ParallelActionScheduler.scala

```diff
@@ -0,0 +1,106 @@
+    Try {
+      var finished: Option[Seq[futureResult]] = None
+      do {
+        val runningActionsCount = pools.map(_._2.running.size).sum
+        logInfo(s"Waiting for an action to finish to continue. Running actions: $runningActionsCount")
+        finished = Option(actionFinishedNotificationQueue.poll(1, TimeUnit.SECONDS))
+          .map { oneAction =>
+            //optimistic attempt to get results for more actions without blocking
+            val bucket = new util.LinkedList[futureResult]()
+            actionFinishedNotificationQueue.drainTo(bucket, 1000)
+            bucket.asScala :+ oneAction
+          }
+      } while (finished.isEmpty)
+      logInfo(s"No more waiting for an action. ${finished.isDefined}")
+      finished.map { rSet =>
+        val poolActionGuids: Map[String, Set[String]] = rSet.map(r => (r._1, r._2.schedulingGuid)).groupBy(_._1).mapValues(v => v.map(_._2).toSet)
+        val newPools = poolActionGuids.foldLeft(pools) { (newPools, kv) =>
+          val newPoolDesc = newPools(kv._1).removeActionGUIDS(kv._2)
+          newPools + (kv._1 -> newPoolDesc)
+        }
+        (new ParallelActionScheduler(newPools, actionFinishedNotificationQueue, poolIntoContext), rSet.map(r => (r._2, r._3)))
+      }
+    }.flatMap {
+      case Some(r) => Success(r)
+      case _ => Failure(new DataFlowException(s"Something when wrong!!! Not sure what happend."))
+    }
+  }
+
+  override def submitAction(poolName: String, action: DataFlowAction[C], entities: DataFlowEntities, flowContext: C): ActionScheduler[C] = {
+    logInfo("Submitting Action: " + poolName + " : " + action.logLabel)
+    val poolDesc = pools.getOrElse(poolName, ExecutionPoolDesc(poolName, 1, Set.empty, ExecutionContext.fromExecutor(Executors.newFixedThreadPool(3))))
+    val ft = Future[futureResult] {
+      logInfo("Executing action " + actionFinishedNotificationQueue.size() + " " + action.logLabel)
+      poolIntoContext(poolName, flowContext)
+      val actionResult = Try(action.performAction(entities, flowContext)).flatten
+      val res = (poolName, action, actionResult)
+      actionFinishedNotificationQueue.offer(res)
+      res
+    }(poolDesc.threadsExecutor)
+    val newPoolDesc = poolDesc.addActionGUID(action.schedulingGuid)
+    new ParallelActionScheduler(pools + (poolName -> newPoolDesc), actionFinishedNotificationQueue, poolIntoContext)
+  }
+
+  override def availableExecutionPools(): Option[Set[String]] = Option(pools.filter(kv => kv._2.running.size < kv._2.maxJobs).keySet).filter(_.nonEmpty)
+
+}
+
+case class ExecutionPoolDesc(poolName: String, maxJobs: Int, running: Set[String], threadsExecutor: ExecutionContextExecutor) {
+
+   def addActionGUID(guid: String): ExecutionPoolDesc = ExecutionPoolDesc(poolName, maxJobs, running + guid, threadsExecutor)
+
+   def removeActionGUIDS(guids: Set[String]): ExecutionPoolDesc = ExecutionPoolDesc(poolName, maxJobs, running &~ guids, threadsExecutor)
+
+}
+
+object ParallelActionScheduler {
```

Comment:
I'm not sure about the poolIntoContext. Maybe the flow context would do the pool assigning. Feels like the user should not be responsible for spark pool assignment implementation.

We could turn FlowContext into a trait and add assign pool function or action prep function etc.

Reply:
I saw that you actually make it easy later on:
```scala
  def sparkMultiJobExecutor(maxJobs: Int, priorityStrategy: priorityStrategy[SparkFlowContext]): DataFlowExecutor[SparkFlowContext] = ParallelDataFlowExecutor[SparkFlowContext](SparkFlowReporter, maxJobs, priorityStrategy)(SparkFlowContext.setPoolIntoContext)
```

But turning the context into a trait could simplify things?